### PR TITLE
fix: preserve inactive cells during masked updates

### LINKED ISSUE #115: 🐛[BUG]: FusedStage masked updates can change inactive systems' cells
### Version

v0.1.0 commit 01c99d5cde6f63d6f662b071a9f408d3bfc12b0a

### On which installation method(s) does this occur?

Source

### Describe the issue

`FusedStage` runs multiple sub-stages on one shared `Batch`. Each sub-stage should update only systems whose `status` matches that stage. For example, in a batch of 2 systems, one system may be in a fixed-cell NVE/NVT validation stage while another system in the same live batch is in a variable-cell FIRE/NPT-style stage.

The masked update helpers restore inactive systems' `positions` and `velocities` after a sub-stage runs, but they do not restore `cell`. As a result, a cell-mutating stage can change every system's cell, including systems that are inactive for that stage.

Expected behavior: inactive systems' `cell` values should be protected during masked updates, just like their positions and velocities.

Actual behavior: inactive systems' `cell` values can change when another system is updated by a cell-mutating stage.

### Minimum reproducible example

```python
from collections import OrderedDict

import torch
from torch import nn

from nvalchemi.data import AtomicData, Batch
from nvalchemi.dynamics import FIREVariableCell
from nvalchemi.models.base import BaseModelMixin, ModelConfig


class ConstantStressModel(nn.Module, BaseModelMixin):
    """Minimal model that provides forces and stress for variable-cell dynamics."""

    def __init__(self) -> None:
        super().__init__()
        self.model_config = ModelConfig(outputs=frozenset({"energy", "forces", "stress"}))

    @property
    def embedding_shapes(self) -> dict[str, tuple[int, ...]]:
        return {}

    def compute_embeddings(self, data: AtomicData, **kwargs: object) -> AtomicData:
        return data

    def forward(self, batch: Batch) -> OrderedDict[str, torch.Tensor]:
        return OrderedDict(
            energy=torch.zeros(batch.num_graphs, 1, device=batch.device),
            forces=torch.zeros_like(batch.positions),
            stress=torch.zeros(batch.num_graphs, 3, 3, device=batch.device),
        )


def make_data(offset: float) -> AtomicData:
    data = AtomicData(
        atomic_numbers=torch.tensor([1], dtype=torch.long),
        positions=torch.tensor([[offset, 0.0, 0.0]], dtype=torch.float32),
        cell=torch.eye(3).unsqueeze(0) * (offset + 1.0),
        pbc=torch.tensor([[True, True, True]]),
    )
    data.forces = torch.zeros(1, 3)
    data.energy = torch.zeros(1, 1)
    data.add_node_property("velocities", torch.zeros(1, 3))
    return data


def main() -> None:
    batch = Batch.from_data_list([make_data(0.0), make_data(1.0)])
    dynamics = FIREVariableCell(model=ConstantStressModel(), dt=1.0, n_steps=1)
    dynamics._ensure_state_initialized(batch)
    dynamics._state.cell_velocity.fill_(0.05)

    mask = torch.tensor([True, False], dtype=torch.bool)
    inactive_before = batch.cell[1].clone()

    # Pre-update is called in FusedStage _step_impl 
    dynamics._masked_pre_update(batch, mask)

    inactive_after = batch.cell[1].clone()
    print(f"inactive cell before:\n{inactive_before}")
    print(f"inactive cell after masked pre_update:\n{inactive_after}")

    if not torch.equal(inactive_before, inactive_after):
        print()
        print("BUG REPRODUCED: inactive system cell changed during masked update.")
        return


if __name__ == "__main__":
    main()
```

### Relevant log output

```shell
inactive cell before:
tensor([[2., 0., 0.],
        [0., 2., 0.],
        [0., 0., 2.]])
inactive cell after masked pre_update:
tensor([[2.1000, 0.1000, 0.1000],
        [0.1000, 2.1000, 0.1000],
        [0.1000, 0.1000, 2.1000]])

BUG REPRODUCED: inactive system cell changed during masked update.
```

### Environment details

```shell
+ Environment location: Source checkout
+ OS: Linux
+ Python/package manager: uv
+ GPU: not required for this reproducer
```

### PR DESCRIPTION
**Pull request by architdatar**
# Protect inactive systems' `cell` tensors during masked dynamics updates.

## Description

`FusedStage` runs multiple sub-stages on one shared live `Batch`. In a mixed workflow, one system may be in a fixed-cell stage such as NVE/NVT while another system is in a variable-cell stage such as FIRE/NPT. The masked update helpers already restore inactive systems' `positions` and `velocities`; this change also restores inactive systems' `cell` values so a cell-mutating sub-stage cannot update every system's cell.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #115 

## Changes Made

- Add `cell` to `BaseDynamics._mutable_fields` so masked update restoration protects inactive system cells.
- Add regression coverage for the split `_masked_pre_update()` path used by `FusedStage`.
- Add regression coverage for the full `masked_update()` helper path used outside the split `FusedStage` step flow.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?
Yes. Added regression tests for both masked update paths: `_masked_pre_update()` for the split `FusedStage` flow and `masked_update()` for the full helper flow. Both verify inactive systems' `cell` values are restored.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

The two regression tests cover independent masking paths:

- `_masked_pre_update()` is the actual split path used by `FusedStage` before the shared model compute.
- `masked_update()` is the full helper path used outside that split `FusedStage` flow.

This is a Python-only toolkit fix. No `nvalchemi-toolkit-ops` changes are required.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: test/dynamics/test_base.py [hunk 1]
```diff
@@ -1434,6 +1434,43 @@ class _ExtendedDynamics(DemoDynamics):
...
+
+        unmasked_cell_before = batch.cell[1].clone()
+        dynamics.masked_update(batch, torch.tensor([True, False]))
+
+        assert torch.equal(batch.cell[1], unmasked_cell_before)
+        assert torch.equal(batch.cell[0], torch.eye(3) + 2.0)
+
 
 # -----------------------------------------------------------------------------
 # Hook frequency gating
```
[the remark this answers is not in the document]
Comment (architdatar) [reply test/dynamics/test_base.py:1493]: Addressed. Added `test_masked_post_update_preserves_unmasked_cell` to cover the split `FusedStage` post-update path with a cell-mutating integrator. The masked update test class now covers `_masked_pre_update()`, `_masked_post_update()`, and full `masked_update()` cell restoration.

**Comment by laserkelvin** [comment]
/ok to test [60e58cd](https://github.com/NVIDIA/nvalchemi-toolkit/pull/116/commits/60e58cdb1c819f55126844910087efe69066c911)

**Review by laserkelvin — APPROVED** [review]
LGTM
